File: app/src/main/java/com/liadpaz/amp/utils/LocalFiles.java
package com.liadpaz.amp.utils;

import android.content.ContentResolver;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.provider.MediaStore.Audio.Media;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.liadpaz.amp.livedatautils.PlaylistsUtil;
import com.liadpaz.amp.livedatautils.QueueUtil;
import com.liadpaz.amp.livedatautils.SongsUtil;
import com.liadpaz.amp.viewmodels.Playlist;
import com.liadpaz.amp.viewmodels.Song;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;

public class LocalFiles {
    private static final String TAG = "AmpApp.LocalFiles";

    private static final String[] PROJECTION = {Media.TITLE, Media._ID, Media.ARTIST, Media.ALBUM, Media.ALBUM_ID};

    private static SharedPreferences musicSharedPreferences;
    private static SharedPreferences playlistsSharedPreferences;

    private static String SORT_DEFAULT = Media.TITLE + " COLLATE NOCASE";

    private static AtomicBoolean isFirstTimeQueue = new AtomicBoolean(true);
    private static AtomicBoolean isFirstTimePosition = new AtomicBoolean(true);

    public static void init(@NonNull Context context) {
        musicSharedPreferences = context.getSharedPreferences("Music.Data", 0);
        playlistsSharedPreferences = context.getSharedPreferences("Music.Playlists", 0);

        ContentResolver contentResolver = context.getContentResolver();

        CompletableFuture.runAsync(() -> PlaylistsUtil.setPlaylists(getPlaylists(contentResolver)));
        CompletableFuture.runAsync(() -> SongsUtil.setSongs(listSongs(contentResolver, SORT_DEFAULT)));

        QueueUtil.observeQueue(songs -> {
            if (!isFirstTimeQueue.getAndSet(false)) {
                ArrayList<Long> songsIdsList = songs.stream().map(song -> song.songId).collect(Collectors.toCollection(ArrayList::new));
                musicSharedPreferences.edit().putString(Constants.PREFERENCES_QUEUE, new Gson().toJson(songsIdsList)).apply();
            }
        });
        QueueUtil.observePosition(queuePosition -> {
            if (!isFirstTimePosition.getAndSet(false)) {
                musicSharedPreferences.edit().putInt(Constants.PREFERENCES_QUEUE_POSITION, queuePosition).apply();
            }
        });

        //        loadQueue(context.getContentResolver());
        // TODO: fix queue saving
    }

    private static void loadQueue(@NonNull ContentResolver contentResolver) {
        List<Long> songsIds = new Gson().fromJson(musicSharedPreferences.getString(Constants.PREFERENCES_QUEUE, "[]"), new TypeToken<ArrayList<Long>>() {}.getType());
        List<Song> songs = listSongs(contentResolver, SORT_DEFAULT, songsIds);
        Collections.sort(songs, Comparator.comparing(song -> songsIds.indexOf(song.songId)));
        QueueUtil.setQueue(songs);
        QueueUtil.setPosition(musicSharedPreferences.getInt(Constants.PREFERENCES_QUEUE_POSITION, -1));
    }

    /**
     * This function returns the music folder path, if it does not exists, it returns the default
     * one.
     *
     * @return The music folder path, if not exists returns the default one.
     */
    @NonNull
    public static String getPath() {
        return musicSharedPreferences.getString(Constants.PREFERENCES_PATH, Constants.DEFAULT_PATH);
    }

    /**
     * This function saves the path of the songs folder to the local shared preferences.
     *
     * @param path The path of the songs folder.
     */
    public static void setPath(@NonNull final String path) {
        musicSharedPreferences.edit().putString(Constants.PREFERENCES_PATH, path).apply();
    }

    /**
     * This function returns the show current song screen on notification click flag.
     *
     * @return true to show current song screen on notification click, otherwise false.
     */
    public static boolean getShowCurrent() {
        return musicSharedPreferences.getBoolean(Constants.PREFERENCES_SHOW_CURRENT, true);
    }

    /**
     * This function sets the show current song screen on notification click flag.
     *
     * @param showCurrent true to show current song screen on notification click flag, otherwise
     *                    false.
     */
    public static void setShowCurrent(final boolean showCurrent) {
        musicSharedPreferences.edit().putBoolean(Constants.PREFERENCES_SHOW_CURRENT, showCurrent).apply();
    }

    /**
     * This function returns the keep screen on when the current song screen is shown flag.
     *
     * @return true to keep screen on when the current song screen is shown, otherwise false.
     */
    public static boolean getScreenOn() {
        return musicSharedPreferences.getBoolean(Constants.PREFERENCES_SCREEN_ON, false);
    }

    /**
     * This function sets the keep screen on when the current song screen is shown flag.
     *
     * @param screenOn true to keep screen on when the current song screen is shown, otherwise
     *                 false.
     */
    public static void setScreenOn(final boolean screenOn) {
        musicSharedPreferences.edit().putBoolean(Constants.PREFERENCES_SCREEN_ON, screenOn).apply();
    }

    @NonNull
    private static ArrayList<Song> listSongs(@NonNull ContentResolver contentResolver, @NonNull String sort) {
        return listSongs(contentResolver, sort, null);
    }

    @NonNull
    private static ArrayList<Song> listSongs(@NonNull ContentResolver contentResolver, @NonNull String sort, @Nullable List<Long> songsIds) {
        long start = System.currentTimeMillis();
        ArrayList<Song> songs = new ArrayList<>();
        //retrieve song info
        try (Cursor musicCursor = contentResolver.query(Media.EXTERNAL_CONTENT_URI, PROJECTION,  Media.DATA + " like ? ", new String[]{"%" + getPath() + "%"}, sort)) {
            //iterate over results if valid
            if (musicCursor != null && musicCursor.moveToFirst()) {
                //get columns
                int titleColumn = musicCursor.getColumnIndex(Media.TITLE);
                int idColumn = musicCursor.getColumnIndex(Media._ID);
                int artistColumn = musicCursor.getColumnIndex(Media.ARTIST);
                int albumColumn = musicCursor.getColumnIndex(Media.ALBUM);
                int albumIdColumn = musicCursor.getColumnIndex(Media.ALBUM_ID);
                //add songs to list
                do {
                    long id = musicCursor.getLong(idColumn);

                    if (songsIds == null || songsIds.contains(id)) {
                        songs.add(new Song(id, musicCursor.getString(titleColumn), musicCursor.getString(artistColumn), musicCursor.getString(albumColumn), musicCursor.getString(albumIdColumn)));
                    }
                } while (musicCursor.moveToNext());
            }
        }
        Log.d(TAG, "listSongs: " + (System.currentTimeMillis() - start));
        return songs;
    }

    @NonNull
    private static Queue<Playlist> getPlaylists(@NonNull ContentResolver contentResolver) {
        long start = System.currentTimeMillis();
        Queue<Playlist> playlists = new ArrayDeque<>();
        playlistsSharedPreferences.getAll().forEach((name, songsIds) -> {
            ArrayList<Long> songsIdsList = new Gson().fromJson(songsIds.toString(), new TypeToken<ArrayList<Long>>() {}.getType());
            ArrayList<Song> songs = listSongs(contentResolver, SORT_DEFAULT, songsIdsList);
            Collections.sort(songs, Comparator.comparing(song -> songsIdsList.indexOf(song.songId)));
            playlists.add(new Playlist(name, songs));
        });
        Log.d(TAG, "getPlaylists: " + (System.currentTimeMillis() - start));
        return playlists;
    }

    public static void setPlaylists(@NonNull Queue<Playlist> playlists) {
        SharedPreferences.Editor editor = playlistsSharedPreferences.edit().clear();
        for (Playlist playlist : playlists) {
            editor.putString(playlist.name, new Gson().toJson((Object)playlist.songs.stream().map(song -> song.songId).collect(Collectors.toCollection(ArrayList::new))));
        }
        editor.apply();
    }

    @NonNull
    public static ArrayList<Song> listSongsByLastAdded(@NonNull ContentResolver contentResolver) {
        return listSongs(contentResolver, Media.DATE_ADDED + " DESC");
    }
}

File: app/src/main/java/com/liadpaz/amp/utils/LocalFiles.kt
package com.liadpaz.amp.utils

import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.provider.MediaStore.Audio
import android.util.Log
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.liadpaz.amp.livedatautils.PlaylistsUtil
import com.liadpaz.amp.livedatautils.QueueUtil
import com.liadpaz.amp.livedatautils.SongsUtil.setSongs
import com.liadpaz.amp.ui.viewmodels.Playlist
import com.liadpaz.amp.ui.viewmodels.Song
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.collections.ArrayList

object LocalFiles {
    private const val TAG = "AmpApp.LocalFiles"

    private val PROJECTION = arrayOf(Audio.Media.TITLE, Audio.Media._ID, Audio.Media.ARTIST, Audio.Media.ALBUM, Audio.Media.ALBUM_ID)
    private const val SORT_DEFAULT = Audio.Media.TITLE + " COLLATE NOCASE"

    private val ART_URI = Uri.parse("content://media/external/audio/albumart")

    private lateinit var musicSharedPreferences: SharedPreferences
    private lateinit var playlistsSharedPreferences: SharedPreferences

    private val isFirstTimeQueue = AtomicBoolean(true)
    private val isFirstTimePosition = AtomicBoolean(true)

    @Suppress("DeferredResultUnused")
    @JvmStatic
    fun init(context: Context) {
        musicSharedPreferences = PreferenceManager.getDefaultSharedPreferences(context.applicationContext)
        playlistsSharedPreferences = context.applicationContext.getSharedPreferences("Music.Playlists", 0)

        val contentResolver = context.contentResolver
        CoroutineScope(Dispatchers.IO).async {
            PlaylistsUtil.setPlaylists(getPlaylists(contentResolver))
            setSongs(listSongs(contentResolver, SORT_DEFAULT))
        }

        QueueUtil.queue.observeForever { songs ->
            if (!isFirstTimeQueue.getAndSet(false)) {
                val songsIdsList = songs.map { song -> song.mediaId }
                musicSharedPreferences.edit().putString(Constants.PREFERENCES_QUEUE, Gson().toJson(songsIdsList)).apply()
            }
        }
        QueueUtil.queuePosition.observeForever { queuePosition ->
            if (!isFirstTimePosition.getAndSet(false)) {
                musicSharedPreferences.edit().putInt(Constants.PREFERENCES_QUEUE_POSITION, queuePosition).apply()
            }
        }

        //        loadQueue(context.getContentResolver());
        // TODO: fix queue saving
    }

    private suspend fun loadQueue(contentResolver: ContentResolver) {
        val songsIds = Gson().fromJson<List<Long>>(musicSharedPreferences.getString(Constants.PREFERENCES_QUEUE, "[]"), object : TypeToken<java.util.ArrayList<String>>() {}.type)
        val songs: List<Song> = listSongs(contentResolver, SORT_DEFAULT, songsIds)
        songs.sortedBy { song: Song -> songsIds.indexOf(song.mediaId) }.also {
            QueueUtil.queue.postValue(ArrayList(it))
            QueueUtil.queuePosition.postValue(musicSharedPreferences.getInt(Constants.PREFERENCES_QUEUE_POSITION, -1))
        }
    }

    @JvmStatic
    val path: String
        get() = musicSharedPreferences.getString(Constants.PREFERENCES_PATH, Constants.DEFAULT_PATH)!!

    @JvmStatic
    val showCurrent: Boolean
        get() = musicSharedPreferences.getBoolean(Constants.PREFERENCES_SHOW_CURRENT, true)

    @JvmStatic
    val screenOn: Boolean
        get() = musicSharedPreferences.getBoolean(Constants.PREFERENCES_SCREEN_ON, false)

    @JvmStatic
    val stopOnTask: Boolean
        get() = musicSharedPreferences.getBoolean(Constants.PREFERENCES_STOP_ON_TASK, false)

    @JvmStatic
    fun listSongs(contentResolver: ContentResolver?, sortBy: String? = null, songsIds: List<Long>? = null): ArrayList<Song> {
        val sort = sortBy ?: SORT_DEFAULT
        contentResolver?.query(Audio.Media.EXTERNAL_CONTENT_URI, PROJECTION, "_data like ? ", arrayOf("%$path%"), sort).use { musicCursor ->
            val songs = arrayListOf<Song>()
            //iterate over results if valid
            if (musicCursor != null && musicCursor.moveToFirst()) {
                //get columns
                val titleColumn = musicCursor.getColumnIndex(Audio.Media.TITLE)
                val idColumn = musicCursor.getColumnIndex(Audio.Media._ID)
                val artistColumn = musicCursor.getColumnIndex(Audio.Media.ARTIST)
                val albumColumn = musicCursor.getColumnIndex(Audio.Media.ALBUM)
                val albumIdColumn = musicCursor.getColumnIndex(Audio.Media.ALBUM_ID)
                //add songs to list
                do {
                    val id = musicCursor.getLong(idColumn)
                    if (songsIds == null || songsIds.contains(id)) {
                        val title = musicCursor.getString(titleColumn)
                        songs.add(Song(id, title, Utilities.getArtistsFromSong(title, musicCursor.getString(artistColumn)), musicCursor.getString(albumColumn), ContentUris.withAppendedId(ART_URI, musicCursor.getString(albumIdColumn).toLong())))
                    }
                } while (musicCursor.moveToNext())
            }
            return songs
        }
    }

    private fun getPlaylists(contentResolver: ContentResolver): Queue<Playlist> {
        val start = System.currentTimeMillis()
        val playlists: Queue<Playlist> = ArrayDeque()
        playlistsSharedPreferences.all.forEach { entry: Map.Entry<String?, Any?> ->
            val songsIdsList = Gson().fromJson<ArrayList<Long>>(entry.value.toString(), object : TypeToken<ArrayList<Long>>() {}.type)
            listSongs(contentResolver, SORT_DEFAULT, songsIdsList).sortedBy { song -> songsIdsList.indexOf(song.mediaId) }.also {
                playlists.add(Playlist(entry.key!!, ArrayList(it)))
            }
        }
        Log.d(TAG, "getPlaylists: " + (System.currentTimeMillis() - start))
        return playlists
    }

    fun setPlaylists(playlists: Queue<Playlist>) {
        val editor = playlistsSharedPreferences.edit().clear()
        for ((name, songs) in playlists) {
            editor.putString(name, Gson().toJson(songs.map { song: Song -> song.mediaId }))
        }
        editor.apply()
    }

    fun listSongsByLastAdded(contentResolver: ContentResolver): java.util.ArrayList<Song> {
        return listSongs(contentResolver, Audio.Media.DATE_ADDED + " DESC")
    }
}